### PR TITLE
#165936552 Restrict similar username and password

### PR DESCRIPTION
**Pull request by bisonlou**
#### What does this PR do?
- Restricts registering a user with a username similar to the password
- Restricts resetting of a password similar to the username

#### How should this be manually tested?
- When registering a user try providing a username that is similar to the password.
- When resetting a password, try entering a password similar to the username.

#### What are the relevant pivotal tracker stories?
- [165936552](https://www.pivotaltracker.com/story/show/165936552)

#### Screenshots (if appropriate)
image

### CODE REVIEW

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -183,6 +190,18 @@ def get_validated_password_or_error(password):
         else:
             return password
 
+def get_unique_username_and_password(data):
+    username = data.get("username", None)
+    password = data.get("password", None)
+
+    if username.lower() == password.lower():
+        raise serializers.ValidationError(
+            "Your username cannot equate to your password"
```
Comment (JamesMudidi) [inline authors/apps/authentication/serializers.py]: Nice work but I think we can have a better error message for the end users.
I think this is not very user-friendly.
I suggest, "Your username and password should not be the same"

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -35,6 +35,12 @@ def create(self, validated_data):
 
         return User.objects.create_user(**validated_data)
 
+    def validate(self, data):
+        # A user should not have a username similar to the password
+
+        return get_unique_username_and_password(data)
+
```
Comment (JamesMudidi) [inline authors/apps/authentication/serializers.py]: Great work here. However, endeavour to use pylint.

**Review by JamesMudidi — CHANGES_REQUESTED** [review]

**Review by llwasampijja — CHANGES_REQUESTED** [review]
Nice work @bisonlou I have tested all the edge cases for signing up and it works perfectly. It would be great if such functionality is also applied to password resetting. Currently, a user can update their password to one which is similar to their username. If possible, try to implement it.

**Comment by nadralia** [comment]
> Nice work @bisonlou I have tested all the edge cases for signing up and it works perfectly. It would be great if such functionality is also applied to password resetting. Currently, a user can update their password to one which is similar to their username. If possible, try to implement it.

**Review by llwasampijja — APPROVED** [review]

**Review by kalsmic — APPROVED** [review]
Nice work @bison, I have tested the functionality and I was not able to set a password that is either similar or contains a user name in it both for registration and password reset.

**Review by JamesMudidi — APPROVED** [review]

**Review by nadralia — APPROVED** [review]
Nice Work @bisonlou

**Pull request merged**